### PR TITLE
new module: GAMMA

### LINKED ISSUE #1531: new module: GAMMA
I think it would be good to have a module for [GAMMA](https://github.com/rastanton/GAMMAl)

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [x] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by sateeshperi**
## PR checklist

Closes #1531  

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/gamma/meta.yml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      description: FASTA file
+      pattern: "*.{fa,fasta}"
+
+output:
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+
```
Comment (rpetit3) [inline modules/gamma/meta.yml:35]:
```suggestion

```

Comment (rpetit3) [reply modules/gamma/meta.yml:35]: @sateeshperi Looks great just need the extra line removed between the outputs

**Review by rpetit3 — COMMENTED** [review]
Looks great, small whitespace fix needed on `meta.yml`

File: modules/gamma/main.nf [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        'quay.io/biocontainers/gamma:2.1--hdfd78af_0' }"
+
+    input:
+    tuple val(meta), path(fasta)
+    path(db)
+
+    output:
+    tuple val(meta), path("*.gamma"), emit: gamma
+    tuple val(meta), path("*.psl")  , emit: psl
+    tuple val(meta), path("*.gff")  , emit: gff      , optional:true
```
Comment (jfy133) [inline modules/gamma/main.nf:19]:
```suggestion
    tuple val(meta), path("*.gff") , optional:true , emit: gff 
```

Not important, but normally we put optional first (I think...)

File: modules/gamma/main.nf [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    GAMMA.py \\
+    $args \\
+    $fasta \\
+    $db \\
+    $prefix
```
Comment (jfy133) [inline modules/gamma/main.nf:33]:
```suggestion
      $args \\
      $fasta \\
      $db \\
      $prefix
```

File: modules/gamma/main.nf [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    conda (params.enable_conda ? "bioconda::gamma=2.1" : null)
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/gamma%3A2.1--hdfd78af_0':
+        'quay.io/biocontainers/gamma:2.1--hdfd78af_0' }"
+
+    input:
+    tuple val(meta), path(fasta)
+    path(db)
+
+    output:
```
Comment (jfy133) [inline modules/gamma/main.nf:16]: Missing `--fasta`

```
-f, --fasta: Writes out a multifasta file of the gene matches
```

File: modules/gamma/meta.yml [hunk 5]
```diff
@@ -0,0 +1,57 @@
+name: "gamma"
+description: Gene Allele Mutation Microbial Assessment
+keywords:
+  - gamma
```
Comment (jfy133) [inline modules/gamma/meta.yml:4]: More keywords as to what the tool is doing please!

File: modules/gamma/meta.yml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - fasta:
+      type: file
+      description: FASTA file
+      pattern: "*.{fa,fasta}"
+  - db:
+      type: file
+      description: FASTA file
```
Comment (jfy133) [inline modules/gamma/meta.yml:26]: Please distinguish from `fasta` input channel

File: modules/gamma/meta.yml [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        e.g. [ id:'test', single_end:false ]
+
+  - versions:
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+
+  - gamma:
+      type: file
+      description: GAMMA file
```
Comment (jfy133) [inline modules/gamma/meta.yml:43]: Would be nice to have more info here of what the file is

**Review by jfy133 — CHANGES_REQUESTED** [review]
Missing an output declaration, and needs more documentation

File: modules/gamma/main.nf (hunk 4 above)
Comment (rastanton) [reply modules/gamma/main.nf:16]: This is an optional argument that can be included as an "args"

Comment (jfy133) [reply modules/gamma/main.nf:16]: Yes sorry I wasn't clear: I meant more that the module would not pick up the output file that comes from that flag.

You don't have to add a varible _within_ the module to execute it, but you must make sure that file does get publishd if a user DOES specify `-f` via $args.

File: modules/gamma/main.nf [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    tuple val(meta), path("*.gamma"), emit: gamma
+    tuple val(meta), path("*.psl")  , emit: psl
+    tuple val(meta), path("*.gff") , optional:true , emit: gff
```
Comment (jfy133) [inline modules/gamma/main.nf:19]: Might be good to check the alignments of `emit` and `optional` as so not to upset certain people ;)

**Review by jfy133 — APPROVED** [review]
Just missing supporting `fasta` output (don't forget to update meta as well) but then you're GTG

File: modules/gamma/main.nf (hunk 4 above)
Comment (sateeshperi) [reply modules/gamma/main.nf:16]: Thank you for clarifying @jfy133 
I have added it to the output and also ran a test to confirm.

**Comment by sateeshperi** [comment]
Thank you for your reviews @jfy133 @rpetit3 @rastanton

**Pull request merged**